File: Leetcode/longest_common_prefix.py
# LeetCode task: solution -> https://leetcode.com/problems/longest-common-prefix/

from typing import List

class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        """Returns the longest common prefix among a list of strings.

        Args:
            strs: A list of strings.

        Returns:
            A string that is the longest common prefix of all the strings in strs.
            If strs is empty or there is no common prefix, returns an empty string.
        """
        if not strs: # if strs is empty
            return "" # return an empty string
        prefix = strs[0] # initialize prefix as the first string in strs
        for i in range(1, len(strs)): # loop through the rest of the strings in strs
            while strs[i].find(prefix) != 0: # while prefix is not a prefix of the current string
                prefix = prefix[:-1] # remove the last character from prefix
                if not prefix: # if prefix becomes empty
                    return "" # return an empty string
        return prefix # return the final prefix

File: Go programming/longest_common_prefix.go
// LeetCode task: solution -> https://leetcode.com/problems/longest-common-prefix/

package main

import (
	"fmt"
	"strings"
)

// longestCommonPrefix returns the longest common prefix among a slice of strings.
// It returns an empty string if the slice is empty or there is no common prefix.
func longestCommonPrefix(strs []string) string {
	if len(strs) == 0 { // if the slice is empty
		return "" // return an empty string
	}
	prefix := strs[0] // initialize prefix as the first string in the slice
	for i := 1; i < len(strs); i++ { // loop through the rest of the strings in the slice
		for !strings.HasPrefix(strs[i], prefix) { // while prefix is not a prefix of the current string
			prefix = prefix[:len(prefix)-1] // remove the last character from prefix
			if len(prefix) == 0 { // if prefix becomes empty
				return "" // return an empty string
			}
		}
	}
	return prefix // return the final prefix
}

func main() {
	strs := []string{"flower", "flow", "flight"} // a sample slice of strings
	result := longestCommonPrefix(strs) // call the function with the sample slice
	fmt.Println(result) // print the result to standard output
	// Output: fl
}
